### PR TITLE
feat: Send notification for user comment

### PR DESCRIPTION
**Pull request by orronai**
- Moderator would receive replies from user
- The moderator would get only one message for user replies
- Added some unittests
- IMPORTANT: Moderator won't get messages between two different replies if he won't comment in the middle

### CODE REVIEW

File: lms/models/solutions.py [hunk 1]
```diff
@@ -6,13 +6,47 @@
...
+            msg = _(
+                '%(checker)s הגיב לך על תרגיל "%(subject)s".',
+                checker=solution.checker.fullname,
+                subject=solution.exercise.subject,
+            )
+            addressee = solution.solver
+
+        if (
+            not solution.comments
+            or solution.comments
```
Comment (yammesicka) [inline lms/models/solutions.py:36]: ?

File: lms/models/solutions.py [hunk 2]
```diff
@@ -6,13 +6,47 @@
...
+            and not solution.comments[-1].commenter == user
```
Comment (yammesicka) [inline lms/models/solutions.py:37]: There's no guarantee they're ordered by time IMHO

File: lms/models/solutions.py [hunk 3]
```diff
@@ -6,13 +6,50 @@
...
+            addressee = solution.checker
+        else:  # solution.checker == user
+            msg = _(
+                '%(checker)s הגיב לך על תרגיל "%(subject)s".',
+                checker=solution.checker.fullname,
+                subject=solution.exercise.subject,
+            )
+            addressee = solution.solver
+        if (
+            not solution.comments
```
Comment (yammesicka) [inline lms/models/solutions.py:34]: This condition should be a function

File: lms/models/solutions.py [hunk 4]
```diff
@@ -6,13 +6,50 @@
...
 
-from lms.lmsdb.models import Solution, SolutionFile
+from lms.lmsdb.models import Comment, Solution, SolutionFile, User
 from lms.lmstests.public.general import tasks as general_tasks
 from lms.lmstests.public.identical_tests import tasks as identical_tests_tasks
 from lms.lmsweb import config, routes
 from lms.models import notifications
 
 
+def send_notification_after_check(user: User, solution: Solution) -> bool:
```
Comment (yammesicka) [inline lms/models/solutions.py:16]: The name of the function is misleading - it's not after check

File: lms/models/solutions.py [hunk 5]
```diff
@@ -6,13 +6,50 @@
...
+    if is_checked:
+        if solution.solver == user:
+            msg = _(
+                '%(solver)s הגיב לך על בדיקת תרגיל "%(subject)s".',
+                solver=solution.solver.fullname,
+                subject=solution.exercise.subject,
+            )
+            addressee = solution.checker
+        else:  # solution.checker == user
+            msg = _(
```
Comment (yammesicka) [inline lms/models/solutions.py:27]: Call the function after the if/else clause. Set a dictionary of the params inside/before the if/else.

File: lms/models/solutions.py [hunk 6]
```diff
@@ -6,13 +6,50 @@
...
+        if (
+            not solution.comments
+            or (
+                solution.comments
+                and not solution.comments.order_by(
+                    Comment.timestamp.desc(),
+                )[0].commenter == user
+            )
+        ):
+            notifications.send(
```
Comment (yammesicka) [inline lms/models/solutions.py:42]: I can't understand the flow. Why we send a notification if `not solution.comments`?

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: lms/models/solutions.py (hunk 6 above)
Comment (orronai) [reply lms/models/solutions.py:42]: First of all we check that the solution has checked.
Then there are two possible situations. There is a possibility that the checker didn't comment at all. And the second one is to check that the last commenter is the checker (if the solver comments) or the solver (if the checker comments).

File: lms/templates/view.html [hunk 7]
```diff
@@ -137,7 +137,7 @@ <h2>{{ _('הערות בודק') }}</h2>
 <script src="{{ url_for('static', filename='prism.js') }}"></script>
 {% if not shared_url %}
 <script src="{{ url_for('static', filename='comments.js') }}"></script>
-{%- if is_manager or config.USERS_COMMENTS %}
+{%- if is_manager or config.USERS_COMMENTS and solution['checker'] %}
```
Comment (yammesicka) [inline lms/templates/view.html:140]: I think we should allow the user to add comments even if the manager still hasn't check the solution

File: lms/models/solutions.py [hunk 8]
```diff
@@ -2,18 +2,64 @@
...
+            return True
+    return False
+
+
+def is_last_to_reply(user: User, solution: Solution) -> bool:
+    return (
+        not solution.comments
+        or (
+            solution.comments
+            and not solution.ordered_comments[0].commenter == user
```
Comment (yammesicka) [inline lms/models/solutions.py:38]: `!=`?

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: lms/tests/test_notifications.py [hunk 9]
```diff
@@ -164,3 +165,54 @@ def test_auto_deletion(self, student_user: User):
...
+        staff_user: User,
+    ):
+        client = conftest.get_logged_user(student_user.username)
+
+        # Marking the solution as checked
+        solutions.mark_as_checked(solution.id, staff_user.id)
+        solution = Solution.get_by_id(solution.id)
+
+        # Sending comments after solution checked
+        comment_response = client.post('/comments', data=json.dumps(dict(
```
Comment (yammesicka) [inline lms/tests/test_notifications.py:182]: Change all the `dict(` to literal dictionaries

**Review by yammesicka — CHANGES_REQUESTED** [review]

**Review by yammesicka — APPROVED** [review]

**Review by yammesicka — APPROVED** [review]

**Pull request merged**